Custom cell and core data

CatalinDavid17 opened the pull request.

Commit:
Added List controller

Commit:
Added Topic Entity + Changed dateField type from ReminderEntity + Changed TopicTableViewController according to the model class

Commit:
Added ReminderController + CustomCell (image -textView - accessory)

Commit:
Removed - ReminderCell

Commit:
Added User entity + model class

Commit:
Added ReminderCell + Keyboard layout adapted + row size dynamically changing

Commit:
Reminder Model - Added new attributes(note, priority)

Commit:
Partial Implementation :
- Added NSFetchedResultsControllerDelegates methods,
- Option to add and store reminder into CoreData by catching EnterKey events
- Must do Update for cells already in(right now that else case is empty)

Commit:
Partial Implementation :
- Added NSFetchedResultsControllerDelegates methods,
- Option to add and store reminder into CoreData by catching EnterKey events
- Must do Update for cells already in(right now that else case is empty)

Commit:
Partial Implementation:
- Update for cell implemented

Commit:
Fixed Crash - When there is only the temporary cell in the tableView, and you try to delete

Commit:
Added Commentary to upadate - Different Approach

Commit:
Partial Implementation:
Modified Reminder Model - Added creationDate attribute
					   - this will keep the order of the created reminders by sorting them on this attribute

Commit:
Adding scheme management files

Commit:
Removed bad indentation + useless code

Commit:
Inject topic from TopicController -> ReminderController via segue

Commit:
Modified ReminderModel - Priority changed to Int type
					  - Warnings storyboard removed
					  - Added predicate to filter reminders on topic

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Model/CDUser+CoreDataProperties.h:
```diff
@@ -0,0 +1,23 @@
+//
```
Remove this class form the model. We will use the Keychain and it is not needed in Core Data anymore.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDTopicTableViewController.m:
```diff
@@ -0,0 +1,223 @@
+	
+}
+
+#pragma mark - Specific action buttons
```
Ar merge "action methods for buttons", din specificatia ta se intelege ca sunt butoane acolo. Schimba textul.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDTopicTableViewController.m:
```diff
@@ -0,0 +1,223 @@
+														 style:UIAlertActionStyleDefault
+													   handler:^(UIAlertAction	*action) {
+														   
+														   NSString *textField = alert.textFields.firstObject.text;
```
Nu e bun numele, tu iei un text aici nu un textfield

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDTopicTableViewController.m:
```diff
@@ -0,0 +1,223 @@
+
+#pragma mark - Helper func for addTapped
+
+- (void) saveTask:(NSString *)topicName {
```
Schimba numele, tu salvezi un topic aici nu un task. saveTopicWithName:

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDTopicTableViewController.m:
```diff
@@ -0,0 +1,223 @@
+	
+	NSError *error;
+	if (![self.managedContext save:&error]) {
+		NSLog(@"An error occured: %@", error);
```
pune mesaje de eroare mai specifice, ca atunci cand o sa iti creasca aplicatia o sa ai o gramada de msaje "An error occured:" si nu sti de ce cod apartiene. Ex "Could not save the Topic objext : Error : %@".
Valabil la toate erorile de acest gen.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDTopicTableViewController.m:
```diff
@@ -0,0 +1,223 @@
+		
+		NSError *error;
+		if (![self.managedContext save:&error]) {
+			NSLog(@"An error occured: %@", error.localizedDescription);
```
error.localizedDescription e mai mult pentru alerte sau ce sa prezinti la user pe UI. Cand logezi o eroare pune obiectul "error" ca vrei sa ai in loguri cat mai multe informatii(codul, domeniul erorii ...).

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Model/CDTopic.h:
```diff
@@ -0,0 +1,24 @@
+
+@interface CDTopic : NSManagedObject
+
+// Insert code here to declare functionality of your managed object subclass
```
Scoate comment-ul

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Model/CDTopic.m:
```diff
@@ -0,0 +1,16 @@
+
+@implementation CDTopic
+
+// Insert code here to add functionality to your managed object subclass
```
Scoate comment-ul

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Model/CDUser.h:
```diff
@@ -0,0 +1,22 @@
+
+@interface CDUser : NSManagedObject
+
+// Insert code here to declare functionality of your managed object subclass
```
Scoate comment-ul

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Model/CDUser.m:
```diff
@@ -0,0 +1,15 @@
+
+@implementation CDUser
+
+// Insert code here to add functionality to your managed object subclass
```
Scoate comment-ul

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -1,35 +1,38 @@
+	self.tableView.rowHeight = UITableViewAutomaticDimension;
+	self.tableView.estimatedRowHeight = 100;
+	
+	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
```
You must remove the observer in the dealoc method. Implement the dealoc method and remove the observers

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -1,35 +1,38 @@
+	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHide:) name:UIKeyboardWillHideNotification object:nil];
 	
 	self.managedContext = ((AppDelegate*)[UIApplication sharedApplication].delegate).managedObjectContext;
 	
```
Error message to generic, be more specific.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +88,229 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+	return CGRectContainsRect(visibleRect, rect);
+	
+}
 
```
Add a pragma mark here for the keyboard specific methods

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +88,229 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+
+- (void)keyboardWillHide:(NSNotification*)aNotification {
+	
+	[UIView beginAnimations:nil context:nil];
```
Varianta asta de cod e invechita. In documentatie scrie "Use of this method is discouraged in iOS 4.0 and later. You should use the block-based animation methods to specify your animations instead." Cauta un exemplu cu varianta pe block sau adapteaza tu codul pe nou a metoda. 

Look at this example http://samwize.com/2015/04/29/how-to-handle-keyboard-hide-slash-show-in-your-view/

Commit:
Partial implementation:
- Removed User from CoreData model
- Added more specific error messages
- Changed saveTask method to saveTopicWithName
- Removed unnecessary commentary
- Fixed other issues regarding to code style

Commit:
Partial implementation:
- Removed User from CoreData model
- Added more specific error messages
- Changed saveTask method to saveTopicWithName
- Removed unnecessary commentary
- Fixed other issues regarding to code style
- Added pragma mark for keyboard specific

Commit:
Partial implementation:
- Removed User from CoreData model
- Added more specific error messages
- Changed saveTask method to saveTopicWithName
- Removed unnecessary commentary
- Fixed other issues regarding to code style
- Added pragma mark for keyboard specific
- Added reloadData to willAppear on TopicController to see the changes to Nr of Items
- Added topic's title to it's Reminder Specific Screen

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
 }
 
-- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
+- (void)scrollToCursorForTextView:(UITextView *)textView {
```
The scrolling logic should go away at some point, because the cell will expand to contain the entire textfield and scrolling won't be necessary anymore.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+
+#pragma mark - UIScrollViewDelegate
 
+- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView {
```
This was just an example, adapt this to work properly for all the cases. Do the changes after the scrolling logic is removed.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+
+- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text {
+	
+	NSIndexPath *indexPath = [NSIndexPath indexPathForRow:self.fetchedResultsController.sections[0].numberOfObjects inSection:0];
```
Give suggestive names to variables to offer the exact context. In this case is lastIndexPathFromFirstSection. And for the cell, "createReminderCell"

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+
+			CDAddReminderCell *cellFromTextView = (CDAddReminderCell *)textView.superview.superview;
+			
+			for (int row = 0; row < ((int)self.fetchedResultsController.sections[0].numberOfObjects); row += 1) {
```
Use fast enumeration.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+			 we can get use of indexPathForSelectedRow? (like in the WWDC video with the gesture events)
+			 */
+
+			CDAddReminderCell *cellFromTextView = (CDAddReminderCell *)textView.superview.superview;
```
Not a good solution, this is a hack and not reliable. When you will change the code and move this to the cell and use delegate methods you will have access to the cell.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+	
+}
+
+#pragma mark - UITextViewDelegate
```
UITextViewDelegate methods should go in the custom cell and be handled there because the textfield is there. To expose actions to the controller you will  implement delegate method for the cell. 

In this way you wont have a large method "+- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text {" and will group things more logically. 
Remind me of this so we can discuss live, because there are more to describe here and just a comment is not enough.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+			CDReminder *reminder = [self.fetchedResultsController objectAtIndexPath:indexPath];
+			cell.textView.text = reminder.taskName;
+		}
+		cell.textView.delegate = self;
```
After changing the logic you should set the cell's delegate here and not the textView delegate.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
-	
-		cell.textLabel.text = reminder.taskName;
-		cell.detailTextLabel.text = reminder.taskHoure;
+		if (indexPath.row < self.fetchedResultsController.sections[indexPath.section].numberOfObjects) {
```
A safer approach here is to check  based on the cell type and in this way the order does not matter. In this case if we move the editing cell from the bottom the code will not work anymore.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -202,58 +343,18 @@ - (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEd
-		NSError *error;
-		if (![self.managedContext save:&error]) {
-			NSLog(@"An error occured: %@", error.localizedDescription);
+		if (indexPath.row < self.fetchedResultsController.sections[indexPath.section].numberOfObjects) {
```
Check after cell type.

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDTopicTableViewController.m:
```diff
@@ -0,0 +1,230 @@
+
+- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
+	
+	UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"SimpleReminderCell"];
```
SimpleReminderCell change name. Topic cell

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
 - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
 	
-	UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"SimpleReminderCell"];
+	CDReminderCell *cell = [tableView dequeueReusableCellWithIdentifier:@"AddCell" forIndexPath:indexPath];
```
Add logic here to show the appropriate cell for the index path. For the last index create a cell with identifier "AddCell" and for the rest create of type "ReminderCellBasic".

Misantrofia commented on ReminderApp-OBJC/ReminderApp-OBJC/Classes/Controller/CDReminderTableViewController.m:
```diff
@@ -129,62 +101,231 @@ - (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)
+	
+	NSLog(@"DidChange \n");
+	
+	[self.tableView beginUpdates];
```
Remove this. After you change something in core data "beginUpdates" and "endUpdates" will be performed by the fetchResultController delegate methods.

Commit:
- Changed TopicCell Identifier
- Changed IndexPath name into textView's delegate method to more relevant ones

Commit:
- Added delegates for CDReminderCell / CDAddReminderCell
- Modified storyboard

Commit:
Huggin priorities changed

Commit:
- Test: Fixed cell height

Commit:
- Partial Implementation:
	- Added custom cells' delegate methods and implementation
	- AwakeFromInsert readded
	- Delete cells only on those of type CDReminderCell
	- Fixed sorting
